File: integration_tests/kotlin_client_tests_real/client-app/src/test/kotlin/dev/graphlink/test/UploadTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GLUpload
import dev.graphlink.test.generated.client.GraphLinkClient
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import kotlin.test.assertTrue

class UploadTest {

    private lateinit var client: GraphLinkClient

    private fun makeFile(vararg bytes: Int, name: String = "test.png"): GLUpload {
        val data = ByteArray(bytes.size) { bytes[it].toByte() }
        return GLUpload(stream = ByteArrayInputStream(data), mimeType = "image/png",
                        length = data.size.toLong(), filename = name)
    }

    @BeforeEach fun setUp() { client = Fixtures.newUploadClient() }

    @Nested inner class UploadOneFile {
        @Test fun `returns true when file is provided`() = runTest {
            val res = client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            assertTrue(res.uploadOneFile)
        }
        @Test fun `returns true for a different userId`() = runTest {
            val res = client.mutations.uploadOneFile("user-42", makeFile(10, 20, 30))
            assertTrue(res.uploadOneFile)
        }
        @Test fun `returns true for a larger file`() = runTest {
            val res = client.mutations.uploadOneFile("user-1",
                GLUpload(stream = ByteArrayInputStream(ByteArray(1024) { (it % 256).toByte() }),
                         mimeType = "application/octet-stream", length = 1024, filename = "big.bin"))
            assertTrue(res.uploadOneFile)
        }
    }

    @Nested inner class UploadFileList {
        @Test fun `returns true for single file in list`() = runTest {
            val res = client.mutations.uploadFileList("user-1", listOf(makeFile(1, 2, 3)))
            assertTrue(res.uploadFileList)
        }
        @Test fun `returns true for two files`() = runTest {
            val res = client.mutations.uploadFileList("user-1",
                listOf(makeFile(1, 2, 3), makeFile(4, 5, 6, name = "second.png")))
            assertTrue(res.uploadFileList)
        }
        @Test fun `returns true for three files`() = runTest {
            val res = client.mutations.uploadFileList("user-2",
                listOf(makeFile(1), makeFile(2), makeFile(3)))
            assertTrue(res.uploadFileList)
        }
    }
}

File: integration_tests/dart_client_tests_real/test/upload_test.dart
import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_uploads.dart';
import 'real_server_adapter.dart';

GLUpload _makeFile(List<int> bytes, {String name = 'test.png'}) =>
    GLUpload.fromBytes(
      Uint8List.fromList(bytes),
      filename: name,
      mimeType: 'image/png',
    );

void main() {
  late GraphLinkClient client;
  setUp(() => client = newUploadClient());

  // ── uploadOneFile ──────────────────────────────────────────────────────────

  group('uploadOneFile', () {
    test('returns true when file is provided', () async {
      final res = await client.mutations.uploadOneFile(
        userId: 'user-1',
        file: _makeFile([1, 2, 3]),
      );
      expect(res.uploadOneFile, isTrue);
    });

    test('returns true for a different userId', () async {
      final res = await client.mutations.uploadOneFile(
        userId: 'user-42',
        file: _makeFile([10, 20, 30]),
      );
      expect(res.uploadOneFile, isTrue);
    });

    test('returns true for a larger file', () async {
      final res = await client.mutations.uploadOneFile(
        userId: 'user-1',
        file: _makeFile(List.generate(1024, (i) => i % 256), name: 'big.bin'),
      );
      expect(res.uploadOneFile, isTrue);
    });
  });

  // ── uploadFileList ─────────────────────────────────────────────────────────

  group('uploadFileList', () {
    test('returns true for a single file in the list', () async {
      final res = await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [
          _makeFile([1, 2, 3])
        ],
      );
      expect(res.uploadFileList, isTrue);
    });

    test('returns true for two files', () async {
      final res = await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [
          _makeFile([1, 2, 3]),
          _makeFile([4, 5, 6], name: 'second.png'),
        ],
      );
      expect(res.uploadFileList, isTrue);
    });

    test('returns true for three files', () async {
      final res = await client.mutations.uploadFileList(
        userId: 'user-2',
        files: [
          _makeFile([1]),
          _makeFile([2]),
          _makeFile([3]),
        ],
      );
      expect(res.uploadFileList, isTrue);
    });
  });
}
